### PR TITLE
feat(frontend): Bind concat_ws expression

### PR DESCRIPTION
**Pull request by kwannoel**
## What's changed and what's your intention?

Implement support for `concat_ws` on frontend.

- Summarize your change (**mandatory**)

  Bind `concat_ws` on frontend. Call cast over function arguments to ensure they are `varchar`.

- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

Part of #2405

### CODE REVIEW

File: src/frontend/src/expr/function_call.rs [hunk 1]
```diff
@@ -115,6 +115,17 @@ impl FunctionCall {
                 }
                 align_types(inputs.iter_mut())
             }
+            ExprType::ConcatWs => {
+                if inputs.len() < 2 {
+                    return Err(ErrorCode::BindError(
+                        "ConcatWs function must contain more than 2 arguments".into(),
```
Comment (xiangjinwu) [inline src/frontend/src/expr/function_call.rs:121]: `CONCAT_WS function must contain at least 2 arguments`

File: src/frontend/src/binder/expr/function.rs [hunk 2]
```diff
@@ -62,6 +62,10 @@ impl Binder {
                     inputs = Self::rewrite_nullif_to_case_when(inputs)?;
                     ExprType::Case
                 }
+                "concat_ws" => {
+                    inputs = Self::rewrite_concat_ws_args_when(inputs)?;
+                    ExprType::ConcatWs
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/function.rs:67]: Just return `ExprType::ConcatWs` here and add casting in `FunctionCall::new`. cast-only changes do not use dedicated `rewrite_xxx` functions.

File: e2e_test/v2/batch/func.slt [hunk 3]
```diff
@@ -69,3 +69,42 @@ drop table t;
...
+statement ok
+create table t (v1 bool, v2 smallint, v3 int, v4 decimal, v5 real, v6 double, v7 varchar, v8 varchar);
+
+statement ok
+insert into t values (true, 1, 2, 3.01, 4, 5.01, 'abc', NULL);
+
+query T
+select concat_ws(v1, v2, v3, v4, v5, v6, v7, v8) from t;
+----
+1true2true3.01true4true5.01trueabc
```
Comment (xiangjinwu) [inline e2e_test/v2/batch/func.slt:107]: These 2 cases are invalid in PostgreSQL. The 0-th arg is required to be string without casting.

**Review by xiangjinwu — COMMENTED** [review]
Mostly lgtm. Just some small adjustments:

**Comment by xiangjinwu** [comment]
Also add a query in `src/frontend/test_runner/tests/testdata/expr.yaml` to improve test coverage.

File: src/frontend/src/expr/function_call.rs [hunk 4]
```diff
@@ -115,6 +115,30 @@ impl FunctionCall {
...
+                    )
+                    .into());
+                }
+
+                // subsequent inputs can be any type, they are cast into varchars with
+                // explicit_cast.
+                inputs = inputs
+                    .into_iter()
+                    .map(|input| input.cast_explicit(DataType::Varchar))
+                    .collect::<Result<Vec<_>>>()?;
```
Comment (xiangjinwu) [inline src/frontend/src/expr/function_call.rs:138]:
```
                inputs = inputs
                    .into_iter()
                    .enumerate()
                    .map(|(i, input)| match i == 0 {
                        // 0-th arg must be string
                        true => input.cast_implicit(DataType::Varchar),
                        // subsequent can be any type
                        false => input.cast_explicit(DataType::Varchar),
                    })
                    .collect::<Result<Vec<_>>>()?;
```
This would also allow the e2e test cases that use literal `null` as separator.

**Review by xiangjinwu — APPROVED** [review]
rest lgtm

File: src/frontend/src/expr/function_call.rs (hunk 4 above)
Comment (kwannoel) [reply src/frontend/src/expr/function_call.rs:138]: Thank you! I was wondering how to patch that.
